pkg/aggregator/internal/limiter/limiter.go

```diff
@@ -0,0 +1,165 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package limiter
+
+import (
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/tagset"
+)
+
+type entry struct {
+	current  int // number of contexts currently in aggregator
+	rejected int // number of rejected samples
+	tags     []string
+}
+
+// Limiter tracks number of contexts based on origin detection metrics
+// and rejects samples if the number goes over the limit.
+//
+// Not thread safe.
+type Limiter struct {
+	key   string
+	tags  []string
+	limit int
+	usage map[string]*entry
+}
+
+// New returns a new instance of limiter.
+//
+// If limit is zero or less the limiter is disabled.
+func New(limit int, key string, tags []string) *Limiter {
+	if limit <= 0 {
+		return nil
+	}
+
+	if !strings.HasSuffix(key, ":") {
+		key += ":"
+	}
+
+	hasKey := false
+	tags = append([]string{}, tags...)
+	for i := range tags {
+		if !strings.HasSuffix(tags[i], ":") {
+			tags[i] += ":"
+		}
+		hasKey = hasKey || key == tags[i]
+	}
+
+	if !hasKey {
+		tags = append(tags, key)
+	}
+
+	return &Limiter{
+		key:   key,
+		tags:  tags,
+		limit: limit,
+		usage: map[string]*entry{},
+	}
+}
+
+func (l *Limiter) identify(tags []string) string {
+	for _, t := range tags {
+		if strings.HasPrefix(t, l.key) {
+			return t
+		}
+	}
+	return ""
+}
+
+func (l *Limiter) extractTags(src []string) []string {
+	dst := make([]string, 0, len(l.tags))
+
+	for _, t := range src {
+		for _, p := range l.tags {
+			if strings.HasPrefix(t, p) {
+				dst = append(dst, t)
+			}
+		}
+	}
+
+	return dst
+}
+
+// Track is called for each new context. Returns true if the sample should be accepted, false
+// otherwise.
+func (l *Limiter) Track(tags []string) bool {
+	if l == nil {
+		return true
+	}
+
+	id := l.identify(tags)
```

Comment:
What should the behavior be for tagsets where there is no `key`? It looks like currently we count all contexts without a sender key into a single bucket and apply the limit to this bucket, is this intentional?

Reply:
The intended behavior is to track all contexts without the key as coming from the same sender. Since our ultimate goal here is to prevent unbounded memory usage in multi-tenant situations, not limiting contexts from senders we could not identify (e.g. from UDP) seems less desirable.